developer-workflow/extension-modules.rst

```diff
@@ -5,13 +5,679 @@
+   might be supplied, overridden or disabled externally.
+
+   Usually, accelerator modules are built as *shared* extension modules,
+   especially if they already have a pure Python implementation.
+
+According to :pep:`399`, *new* extension modules MUST provide a working and
+tested pure Python implementation, unless a special dispensation is given.
+Please ask the :github:`Steering Council <python/steering-council>` if such
+dispensation is needed.
+
+Adding an extension module to CPython
+=====================================
+
+Assume that the standard library contains a pure Python module :mod:`!foo`
+together with the following :func:`!foo.greet` function:
+
+.. code-block:: python
+   :caption: Lib/foo.py
+
+   def greet():
+       return "Hello World!"
+
+Instead of using the Python implementation of :func:`!foo.greet`, we want to
+use its corresponding C implementation exposed in some :mod:`!_foo` module
+written in C. Ideally, we want to modify :cpy-file:`!Lib/foo.py` as follows:
+
+.. code-block:: python
+   :caption: Lib/foo.py
+
+   try:
+       # use the C implementation if possible
+       from _foo import greet
+   except ImportError:
+       # fallback to the pure Python implementation
+       def greet():
+           return "Hello World!"
+
+.. note::
+
+   Accelerator modules should *never* be imported directly, whence the
+   convention is to mark them as private implementation details with the
+   underscore prefix (namely, :mod:`!_foo` in this example).
+
+In order to incorporate the accelerator module, we need to determine:
+
+- where to place the extension module source code in the CPython project tree,
+- which files to modify in order to compile the CPython project, and
+- which ``Makefile`` rules to invoke at the end.
+
+Updating the CPython project tree
+---------------------------------
+
+Usually, accelerator modules are added in the :cpy-file:`Modules` directory of
+the CPython project. If more than one file is needed for the extension module,
+it is more convenient to create a sub-directory in :cpy-file:`Modules`.
+
+For our extension module :mod:`!_foo`, we consider the following working tree:
+
+- :ref:`Modules/_foo/_foomodule.c` --- the extension module implementation.
+- :ref:`Modules/_foo/helper.h` --- the extension helpers declarations.
+- :ref:`Modules/_foo/helper.c` --- the extension helpers implementations.
+
+By convention, the source file containing the extension module implementation
+is called ``<NAME>module.c``, where ``<NAME>`` is the name of the module that
+will be later imported (in our case :mod:`!_foo`). In addition, the directory
+containing the implementation should also be named similarly.
+
+One could imagine having more files, or no helper files at all. Here,
+we wanted to illustrate a simple example without making it too trivial. If
+the extension module does not require additional files, it may directly be
+placed in :cpy-file:`Modules` as ``Modules/_foomodule.c``.
```

Comment:
```suggestion
Usually, accelerator modules are added in the :cpy-file:`Modules` directory of
the CPython project. If more than one file is needed for the extension module,
it is more convenient to create a sub-directory in :cpy-file:`Modules`.

In the simplest example where the extension module is one file, it may be
placed in :cpy-file:`Modules` as ``Modules/_foomodule.c``.

For a non-trivial example of the extension module :mod:`!_foo`, we consider
the following working tree:

- :ref:`Modules/_foo/_foomodule.c` --- the extension module implementation.
- :ref:`Modules/_foo/helper.h` --- the extension helpers declarations.
- :ref:`Modules/_foo/helper.c` --- the extension helpers implementations.

By convention, the source file containing the extension module implementation
is called ``<NAME>module.c``, where ``<NAME>`` is the name of the module that
will be later imported (in our case :mod:`!_foo`). In addition, the directory
containing the implementation should also be named similarly.
```